/**
 * e8yes demo web server.
 *
 * <p>Copyright (C) 2020 Chifeng Wen {daviesx66@gmail.com}
 *
 * <p>This program is free software: you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * <p>You should have received a copy of the GNU General Public License along with this program. If
 * not, see <http://www.gnu.org/licenses/>.
 */

syntax = "proto3";

package e8;

import "procedural_shape.proto";

/**
 * @brief The ProceduralObjectProto message A procedural object usually has geometry defined by 
 * parameters other than primitive meshes and appearance given by data other than texture images.
 */
message ProceduralObjectProto {
    /**
     * @brief The ProceduralObjectType enum All possible types of procedural processes.
     */
    enum ProceduralObjectType {
        INVALID = 0;
        PLANE = 1;
        BOX = 2;
        SPHERE = 3;
    }

    // ID of this procedural object.
    int64 id = 1;

    // A descriptive human readable name of the object.
    string name = 2;

    // The type of procedural process to use.
    ProceduralObjectType type = 3;

    // Indicates if the object is movable.
    bool movable = 4;

    // Source data for a procedural generator to produce geometries.
    repeated ProceduralShapeProto shapes = 5;
}
